blockwork/config/config.py

```diff
@@ -1,6 +1,90 @@
+    def __init__(self, ctx: "Context", site: base.Site, project: base.Project, target: base.Element) -> None:
+        self.ctx = ctx
+        self.site = site
+        self.project = project
+        self.target = target
+
+    def depth_first_elements(self, element: base.Element) -> Iterable[base.Element]:
+        'Recures elements and yields depths first'
+        for sub_element in element.iter_sub_elements():
+            yield from self.depth_first_elements(sub_element)
+        yield element
+
+    def resolve(self):
+        'Resolves paths in the config'
+        linker = Linker(self)
+        for element in self.depth_first_elements(self.target):
+            if isinstance(element, base.Transform):
+                linker.link_outputs(element)
+
+        for element in self.depth_first_elements(self.target):
+            linker.link_inputs(element)
+
+class Linker:
+    '''
+    This class is responsible for finalising input and output paths
+    and linking them together.
+
+    @ed.kotarski: The next extension to this will be to use that
+    information to create dependencies between transforms to determine 
+    the order they need to be run.
+    '''
+    def __init__(self, config: Config):
+        self.config = config
+        self.pairs: dict[tuple[str, str], Path] = {}
+
+    def link_inputs(self, element: base.Element):
+        block = cast(base.ElementContext, element._context).block
+        block_root = self.config.ctx.host_root / self.config.project.blocks[block]
+
+        @overload
+        def resolve(paths: str) -> str: ...
+        @overload
+        def resolve(paths: list[str]) -> list[str]: ...
+        def resolve(paths):
+            if (is_single := isinstance(paths, str)):
+                paths = [paths]
+
+            resolved_paths: list[str] = []
+            for path in paths:
+                pair = (block, path)
+                if pair in self.pairs:
+                    full_path = self.pairs[pair]
+                else:
+                    full_path = (block_root / path)
+                resolved_paths.append(full_path.as_posix())
+            return resolved_paths[0] if is_single else resolved_paths
+
+        element.resolve_input_paths(resolve)
+
+    def link_outputs(self, element: base.Transform):
+        block = cast(base.ElementContext, element._context).block
+        block_scratch = self.config.ctx.host_scratch / self.config.project.blocks[block]
+
+        @overload
+        def resolve(paths: str) -> str: ...
+        @overload
+        def resolve(paths: list[str]) -> list[str]: ...
+        def resolve(paths):
+            if (is_single := isinstance(paths, str)):
+                paths = [paths]
+
+            resolved_paths: list[str] = []
+            for path in paths:
+                full_path = block_scratch / path
+                self.pairs[(block, path)] = full_path
+                resolved_paths.append(full_path.as_posix())
```

Comment:
Again could yield?

Reply:
Discussed - would make more complex for caller